src/main/java/sirius/pasta/noodle/compiler/NoodleCompiler.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Made with all the love in the world
+ * by scireum in Remshalden, Germany
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.pasta.noodle.compiler;
+
+import parsii.tokenizer.LookaheadReader;
+import sirius.pasta.noodle.Callable;
+import sirius.pasta.noodle.ConstantCall;
+import sirius.pasta.noodle.NLSCall;
+import sirius.pasta.noodle.OpCode;
+import sirius.pasta.noodle.ReturnVariableCall;
+import sirius.pasta.noodle.compiler.ir.Constant;
+import sirius.pasta.noodle.compiler.ir.IntrinsicCall;
+import sirius.pasta.noodle.compiler.ir.Node;
+import sirius.pasta.noodle.compiler.ir.PushTemporary;
+import sirius.pasta.noodle.compiler.ir.ReturnNode;
+
+import java.util.Optional;
+
+/**
+ * Uses the {@link Parser} to compile either a {@link #compileScript() script} or an
+ * {@link #compileExpression(boolean) expression} into a {@link Callable}.
+ */
+public class NoodleCompiler {
+
+    private final Parser parser;
+    private final CompilationContext context;
+
+    /**
+     * Creates a new instance which operates on the given input and uses the given context.
+     *
+     * @param context the context used during compilation
+     */
+    public NoodleCompiler(CompilationContext context) {
+        this(context, context.getSourceCodeInfo().createReader());
+    }
+
+    /**
+     * Creates a new instance which operates on the given input and uses the given context.
+     *
+     * @param context the context used during compilation
+     * @param reader  the reader pointing at the expression to parse
+     */
+    public NoodleCompiler(CompilationContext context, LookaheadReader reader) {
+        this.parser = new Parser(context, reader);
+        this.context = context;
+    }
+
+    /**
+     * Compiles a whole script (which can be a block of statements).
+     * <p>
+     * This will essentially instruct the parser to parse a block of statements and then run an optimization
+     * step in case the script can be replaced by a simpligied {@link Callable}.
+     *
+     * @return the compiled script. Note however, the {@link CompilationContext} has to be checked for errors. We
+     * do not raise an exception here, as e.g. <tt>Tagilatelle</tt> might which to compile a complete template and
+     * then report all detected errors.
+     */
+    public Callable compileScript() {
+        Node ir = parseBlock();
```

Comment:
what does `ir` stand for? Could we use a more readable variable name? Also applies to other methods below